{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyMTqvfaqbX2lISXEcGIhlT4",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/LF7-AI/Coding-for-Computational-Biologist/blob/main/MSA_Notebook.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#1.Perform MSA with nucleic acid sequence of your interest:\n",
        "\n",
        "#Step-1: Install required tools\n",
        "!pip install biopython\n",
        "!apt-get install -y clustalo\n",
        "\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "rbO8voDUWMPD",
        "outputId": "2f30dee1-1cbc-46d6-b581-a3f096c1104a"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting biopython\n",
            "  Downloading biopython-1.85-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (13 kB)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.12/dist-packages (from biopython) (2.0.2)\n",
            "Downloading biopython-1.85-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.3 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.3/3.3 MB\u001b[0m \u001b[31m20.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: biopython\n",
            "Successfully installed biopython-1.85\n",
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "The following additional packages will be installed:\n",
            "  libargtable2-0\n",
            "The following NEW packages will be installed:\n",
            "  clustalo libargtable2-0\n",
            "0 upgraded, 2 newly installed, 0 to remove and 35 not upgraded.\n",
            "Need to get 273 kB of archives.\n",
            "After this operation, 694 kB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libargtable2-0 amd64 13-1.1 [14.1 kB]\n",
            "Get:2 http://archive.ubuntu.com/ubuntu jammy/universe amd64 clustalo amd64 1.2.4-7 [259 kB]\n",
            "Fetched 273 kB in 1s (294 kB/s)\n",
            "Selecting previously unselected package libargtable2-0.\n",
            "(Reading database ... 126374 files and directories currently installed.)\n",
            "Preparing to unpack .../libargtable2-0_13-1.1_amd64.deb ...\n",
            "Unpacking libargtable2-0 (13-1.1) ...\n",
            "Selecting previously unselected package clustalo.\n",
            "Preparing to unpack .../clustalo_1.2.4-7_amd64.deb ...\n",
            "Unpacking clustalo (1.2.4-7) ...\n",
            "Setting up libargtable2-0 (13-1.1) ...\n",
            "Setting up clustalo (1.2.4-7) ...\n",
            "Processing triggers for libc-bin (2.35-0ubuntu3.8) ...\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_loader.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_opencl.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtcm.so.1 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libumf.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtcm_debug.so.1 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_0.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero_v2.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbb.so.12 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc_proxy.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_5.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libhwloc.so.15 is not a symbolic link\n",
            "\n",
            "Processing triggers for man-db (2.10.2-1) ...\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#Step-2: Import Modules:\n",
        "from Bio import SeqIO\n",
        "from Bio.Seq import Seq\n",
        "from Bio.SeqRecord import SeqRecord\n",
        "from collections import Counter\n",
        "import subprocess\n"
      ],
      "metadata": {
        "id": "xGpuPMTlWUqr"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Step-3: Define sequences and write to FASTA File\n",
        "sequences=[\n",
        "    SeqRecord(Seq(\"ATTAAGGGCCCCCACGTGCTACCATG\"), id= \"seq1\", description= \"\"),\n",
        "    SeqRecord(Seq(\"AAGGACCCATCACTGTGCTACTCCAGGG\"), id= \"seq2\", description= \"\"),\n",
        "    SeqRecord(Seq(\"ATGCGTACGTACGTACGTAGCTAGCTAGCTAGCTAGC\"), id= \"seq3\", description= \"\")\n",
        "\n",
        "]\n",
        "na_fasta_file = \"na_sequences.fasta\"\n",
        "SeqIO.write(sequences, na_fasta_file, \"fasta\")\n",
        "\n",
        "with open(na_fasta_file, 'r') as f:\n",
        "    nucleic_acid_seqs = f.read()\n",
        "print(nucleic_acid_seqs)\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "SVPGx68KWbPJ",
        "outputId": "cc197672-2703-434f-801d-3e4fd33b9afe"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            ">seq1\n",
            "ATTAAGGGCCCCCACGTGCTACCATG\n",
            ">seq2\n",
            "AAGGACCCATCACTGTGCTACTCCAGGG\n",
            ">seq3\n",
            "ATGCGTACGTACGTACGTAGCTAGCTAGCTAGCTAGC\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#Step-4: Run Clustal Omega alignment for MSA\n",
        "aligned_na_file= \"aligned_na_sequences.fasta\"\n",
        "command= f\"clustalo -i {na_fasta_file} -o {aligned_na_file} --outfmt fasta\"\n",
        "subprocess.run(command, shell=True, check=True)\n",
        "\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ZdPcvM_UW9F6",
        "outputId": "eab2df20-87aa-403f-abe6-73807004a5e6"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "CompletedProcess(args='clustalo -i na_sequences.fasta -o aligned_na_sequences.fasta --outfmt fasta', returncode=0)"
            ]
          },
          "metadata": {},
          "execution_count": 6
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#Step 5: Read and display the aligned sequences\n",
        "from Bio import AlignIO\n",
        "alignment = AlignIO.read(aligned_na_file, \"fasta\")\n",
        "print(\"Aligned Sequences:\")\n",
        "for record in alignment:\n",
        "    print(f\"{record.id}: {record.seq}\")\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "3tbppODOXea9",
        "outputId": "aa3210e7-f0df-435a-a0c7-1dbb2ee41a98"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Aligned Sequences:\n",
            "seq1: ATTAAGGGCCCCCACG--TGCTACCATG------------\n",
            "seq2: ---AAGGACCCATCACTGTGCTACTCCAGGG---------\n",
            "seq3: ---ATGCGTACGTACGTACGTAGCTAGCTAGCTAGCTAGC\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#2.Perform MSA with protein sequence of your interest:\n",
        "# I am not loading or importing libraries as they are already been called from above promt.\n",
        "\n",
        "#Step 1: Define sequences and write to FASTA file:\n",
        "\n",
        "sequences=[\n",
        "         SeqRecord(Seq(\"MKTAYIAKQRISFVKSHFSSRRQDILDLWQY\"), id= \"seq1\", description= \"\"),\n",
        "          SeqRecord(Seq(\"MKTAYIAKQRISGTTTSRRRQDILWMKTYY\"), id= \"seq2\", description= \"\"),\n",
        "         SeqRecord(Seq(\"MKTAYIAKQRISGTTAAADQTSRRRQDILWMKTYY\"), id= \"seq3\", description= \"\")\n",
        "\n",
        "]\n",
        "fasta_file = \"sequences.fasta\"\n",
        "SeqIO.write(sequences, fasta_file, \"fasta\")\n",
        "\n",
        "with open(fasta_file, 'r') as f:\n",
        "        file_content = f.read()\n",
        "print(file_content)\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "a4b6xee5XqHV",
        "outputId": "70e38fd3-14c7-49d2-dffa-3608bebfb25e"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            ">seq1\n",
            "MKTAYIAKQRISFVKSHFSSRRQDILDLWQY\n",
            ">seq2\n",
            "MKTAYIAKQRISGTTTSRRRQDILWMKTYY\n",
            ">seq3\n",
            "MKTAYIAKQRISGTTAAADQTSRRRQDILWMKTYY\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#Step 2: Run Clustal Omega alignment:\n",
        "\n",
        "aligned_file = \"aligned_sequences.fasta\"\n",
        "command= f\"clustalo -i {fasta_file} -o {aligned_file} --outfmt fasta\"\n",
        "subprocess.run(command, shell=True, check=True)\n",
        "\n",
        "#Step 3: Read and display the aligned sequences:\n",
        "\n",
        "from Bio import AlignIO\n",
        "alignment = AlignIO.read(aligned_file, \"fasta\")\n",
        "print(\"Aligned Sequences:\")\n",
        "for record in alignment:\n",
        "    print(f\"{record.id}: {record.seq}\")\n",
        "\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "6fhfeHN3YdU_",
        "outputId": "9a5fb760-de12-47eb-9395-897e19b4a94e"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Aligned Sequences:\n",
            "seq1: MKTAYIAKQRISFVKSH---FSSRRQDILDLWQY-\n",
            "seq2: MKTAYIAKQRISGTT-----TSRRRQDILWMKTYY\n",
            "seq3: MKTAYIAKQRISGTTAAADQTSRRRQDILWMKTYY\n"
          ]
        }
      ]
    }
  ]
}